File: sdk/typescript/src/transport.ts
/** One authenticated fetch transport, with bounded requests and explicit cancellation. */
export enum Method {GET = 'GET', POST = 'POST', PUT = 'PUT', DELETE = 'DELETE'}
export enum MediaType {JSON = 'application/json', BINARY = 'application/octet-stream'}

type QueryValue = string | number | boolean | readonly (string | number | boolean)[] | null | undefined;
export interface CallOptions {
  signal?: AbortSignal | undefined;
  /** Replaces the transport's default deadline for this call. */
  timeoutMs?: number | undefined;
}
export interface TransportOptions {timeoutMs?: number}
interface RequestOptions extends CallOptions {
  parameters?: Record<string, string>;
  query?: Record<string, QueryValue>;
  body?: string | Uint8Array;
  contentType?: MediaType;
  accept?: MediaType;
}

export class HttpError extends Error {
  override readonly name = 'HttpError';
  constructor(readonly status: number, readonly body: string) {
    super(`HTTP ${status}: ${body}`);
  }
}

export class NetworkError extends Error {
  override readonly name = 'NetworkError';
  constructor(cause: unknown) {
    super(cause instanceof Error ? cause.message : 'Gateway connection failed', {cause});
  }
}

/**
 * `AbortSignal.any` where it exists; WKWebView before macOS 14.4 (the desktop
 * app supports 14.0) and older runtimes need the linking done by hand.
 */
function checkedTimeout(timeout: number): number {
  if (!Number.isInteger(timeout) || timeout <= 0 || timeout > 2 ** 31 - 1) {
    throw new TypeError('Timeout must be a positive integer no larger than 2147483647 milliseconds');
  }
  return timeout;
}

function anySignal(signals: AbortSignal[]): AbortSignal {
  if (typeof AbortSignal.any === 'function') return AbortSignal.any(signals);
  const linked = new AbortController();
  const unlink = (): void => {for (const signal of signals) signal.removeEventListener('abort', abort);};
  function abort(this: AbortSignal): void {unlink(); linked.abort(this.reason);}
  for (const signal of signals) {
    if (signal.aborted) {linked.abort(signal.reason); return linked.signal;}
  }
  for (const signal of signals) signal.addEventListener('abort', abort, {once: true});
  return linked.signal;
}

export class Transport {
  readonly #url: string;
  readonly #token: string;
  readonly #timeoutMs: number;
  readonly #closed = new AbortController();

  constructor(url: string, token: string, options: TransportOptions = {}) {
    const parsed = new URL(url);
    if (!['http:', 'https:'].includes(parsed.protocol) || parsed.username || parsed.password || parsed.search || parsed.hash) {
      throw new TypeError('Gateway URL must be HTTP(S), without credentials, query or fragment');
    }
    if (!token || /[\r\n]/.test(token)) throw new TypeError('Gateway bearer token is required without line breaks');
    const timeout = checkedTimeout(options.timeoutMs ?? 30_000);
    this.#url = parsed.href.replace(/\/$/, '');
    this.#token = token;
    this.#timeoutMs = timeout;
  }

  /** The default per-request deadline, in milliseconds. */
  get timeoutMs(): number {return this.#timeoutMs;}

  close(): void {this.#closed.abort();}

  async request(method: Method, path: string, options: RequestOptions = {}): Promise<Uint8Array> {
    if (this.#closed.signal.aborted) throw new Error('SDK client is closed');
    if (!path.startsWith('/') || /[?#]/.test(path)) throw new TypeError('Operation path must be absolute without query or fragment');
    for (const [name, value] of Object.entries(options.parameters ?? {})) {
      // WHATWG URLs normalize even percent-encoded dot segments before fetch.
      if (!value || value === '.' || value === '..') throw new TypeError('Invalid path identifier');
      path = path.replaceAll(`{${name}}`, encodeURIComponent(value).replaceAll('.', '%2E'));
    }
    if (/[{}]/.test(path)) throw new TypeError('Operation has unresolved path parameters');
    const query = new URLSearchParams();
    for (const [key, value] of Object.entries(options.query ?? {})) {
      if (value !== undefined && value !== null) query.set(key, Array.isArray(value) ? value.join(',') : String(value));
    }
    const headers = new Headers({Authorization: `Bearer ${this.#token}`, Accept: options.accept ?? MediaType.JSON});
    if (options.body !== undefined) headers.set('Content-Type', options.contentType ?? MediaType.JSON);
    const deadline = options.timeoutMs === undefined ? this.#timeoutMs : checkedTimeout(options.timeoutMs);
    const signals = [this.#closed.signal, AbortSignal.timeout(deadline)];
    if (options.signal) signals.push(options.signal);
    const signal = anySignal(signals);
    let response: Response;
    let payload: Uint8Array;
    try {
      response = await fetch(this.#url + path + (query.size ? `?${query}` : ''), {
        method, headers, redirect: 'manual', signal,
        ...(options.body === undefined ? {} : {body: typeof options.body === 'string' ? options.body : new Uint8Array(options.body)}),
      });
      payload = new Uint8Array(await response.arrayBuffer());
    } catch (error) {
      if (signal.aborted) throw error;
      throw new NetworkError(error);
    }
    if (!response.ok) throw new HttpError(response.status, new TextDecoder().decode(payload));
    return payload;
  }
}

File: sdk/rust/src/transport.rs
//! Shared HTTP connection pool. Dropping a request future cancels that request.

use std::time::Duration;

use reqwest::header::{HeaderValue, ACCEPT, AUTHORIZATION, CONTENT_TYPE};
use reqwest::{Client, Method, Url};

use crate::{Error, Result};

#[derive(Debug, Clone, Copy, Default)]
pub enum MediaType {
    #[default]
    Json,
    Binary,
}

impl MediaType {
    fn as_str(self) -> &'static str {
        match self {
            Self::Json => "application/json",
            Self::Binary => "application/octet-stream",
        }
    }
}

/// An optional deadline for one request, including reading the response body.
#[derive(Debug, Clone, Copy, Default)]
pub struct CallOptions {
    pub timeout: Option<Duration>,
}

#[derive(Default)]
pub struct Request<'a> {
    pub parameters: &'a [(&'a str, &'a str)],
    pub query: &'a [(&'a str, String)],
    pub body: Option<Vec<u8>>,
    pub content_type: MediaType,
    pub accept: MediaType,
    pub options: CallOptions,
}

/// Clones share a connection pool; no local discovery, subprocesses or UDS access.
///
/// ```no_run
/// use capsem_sdk::transport::{Transport, Request};
/// use std::time::Duration;
/// # async fn example() -> capsem_sdk::Result<()> {
/// let client = Transport::new("http://127.0.0.1:19222", "token", Duration::from_secs(30))?;
/// let bytes = client.request(reqwest::Method::GET, "/status", Request::default()).await?;
/// let info: capsem_sdk::models::HypervisorInfo = serde_json::from_slice(&bytes)?;
/// assert_eq!(info.service, capsem_sdk::models::ServiceAvailability::Running);
/// # Ok(())
/// # }
/// ```
#[derive(Debug, Clone)]
pub struct Transport {
    base: Url,
    client: Client,
    authorization: HeaderValue,
}

impl Transport {
    pub fn new(url: &str, token: &str, timeout: Duration) -> Result<Self> {
        let base = Url::parse(url).map_err(|_| Error::InvalidInput("gateway URL is invalid"))?;
        if !matches!(base.scheme(), "http" | "https")
            || !base.username().is_empty()
            || base.password().is_some()
            || base.query().is_some()
            || base.fragment().is_some()
        {
            return Err(Error::InvalidInput(
                "gateway URL must be HTTP(S) without credentials, query or fragment",
            ));
        }
        if token.is_empty() || timeout.is_zero() {
            return Err(Error::InvalidInput("bearer token and positive timeout are required"));
        }
        let mut authorization = HeaderValue::from_str(&format!("Bearer {token}"))
            .map_err(|_| Error::InvalidInput("bearer token contains invalid header characters"))?;
        authorization.set_sensitive(true);
        let client = Client::builder()
            .timeout(timeout)
            .redirect(reqwest::redirect::Policy::none())
            .retry(reqwest::retry::never())
            .build()?;
        Ok(Self {
            base,
            client,
            authorization,
        })
    }

    pub async fn request(&self, method: Method, path: &str, request: Request<'_>) -> Result<Vec<u8>> {
        if !path.starts_with('/') || path.contains(['?', '#']) {
            return Err(Error::InvalidInput(
                "operation path must be absolute without query or fragment",
            ));
        }
        let mut url = self.base.clone();
        {
            let mut segments = url.path_segments_mut().expect("HTTP URL supports path segments");
            segments.pop_if_empty();
            for segment in path[1..].split('/') {
                let value = if let Some(name) = segment.strip_prefix('{').and_then(|s| s.strip_suffix('}')) {
                    request
                        .parameters
                        .iter()
                        .find(|(key, _)| *key == name)
                        .map(|(_, value)| *value)
                        .ok_or(Error::InvalidInput("missing path parameter"))?
                } else {
                    if segment.contains(['{', '}']) {
                        return Err(Error::InvalidInput("unresolved path parameter"));
                    }
                    segment
                };
                // URL libraries normalize dot-only segments, including escaped ones.
                if value.is_empty() || matches!(value, "." | "..") {
                    return Err(Error::InvalidInput("invalid path identifier"));
                }
                segments.push(value);
            }
        }
        if request.options.timeout.is_some_and(|timeout| timeout.is_zero()) {
            return Err(Error::InvalidInput("request timeout must be positive"));
        }
        let mut outgoing = self
            .client
            .request(method, url)
            .query(request.query)
            .header(AUTHORIZATION, self.authorization.clone())
            .header(ACCEPT, request.accept.as_str());
        if let Some(body) = request.body {
            outgoing = outgoing.header(CONTENT_TYPE, request.content_type.as_str()).body(body);
        }
        if let Some(timeout) = request.options.timeout {
            outgoing = outgoing.timeout(timeout);
        }
        let response = outgoing.send().await?;
        let status = response.status();
        let body = response.bytes().await?.to_vec();
        if !status.is_success() {
            return Err(Error::Http {
                status: status.as_u16(),
                body,
            });
        }
        Ok(body)
    }
}

#[cfg(test)]
mod tests;
